src/com/spike/ErrorReporter.java

```diff
@@ -36,21 +36,32 @@ public void resetErrors() {
     hadErrors = false;
   }
 
-  protected String format(int line, String message) {
-    return String.format("Error at line %d:\n\t%s", line, message);
+  protected String format(int line, int column, String message) {
+    String s = String.format("Error at line %d, col %d\n\t", line, column);
+    return s + lines[line - 1] + "\n\t" +
+      " ".repeat(column - 1) + "^\n\t" +
+      "-".repeat(column - 1) + "|\n\n" +
+      message;
   }
 
   ErrorReporter setSource(String source) {
     lines = source.split("\n");
     return this;
   }
 
-  void report(int line, String message) {
+  void report(String message) {
+    writer.println("Error: " + message);
+  }
+
+  void report(int line, int column, String message) {
     hadErrors = true;
-    writer.println(format(line, message));
+    writer.println(format(line, column, message));
   }
   
   protected String format(Token token, String message) {
+    if (token.line == -1 && token.column == -1)
```

Comment:
To provide proper type checking for `print` builtin I had to manually create AST for it and after provide implementation in Interpreter. While creating AST I had to create several new tokens, but as they hadn't line an number I also had to adjust error reporting, so if we have `print(123)`, type checker will raise an error saying `'print' expects 'String' as argument` but error reporter doesn't know the line and column where the problem is, because `print` is provided internally. Therefore I just print `Error in builtin 'print' + message` without line/column info.